### PR TITLE
int8 and uint8 support in Pad and other ops

### PR DESCRIPTION
**Pull request by igor-yusupov**
[Change description](https://github.com/robertknight/rten/pull/387#issuecomment-2420349780)

### CODE REVIEW

File: .DS_Store
Comment (robertknight) [inline .DS_Store:1]: Can you remove this macOS-specific `.DS_Store` file. I suggest creating a [global `.gitignore`](https://sebastiandedeyne.com/setting-up-a-global-gitignore-file/) for this.

File: src/ops/convert.rs [hunk 1]
```diff
@@ -8,14 +8,27 @@ fn cast(pool: &TensorPool, input: Input, dtype: DataType) -> Result<Output, OpEr
...
+            Input::UInt8Tensor(t) => Ok(t.map_in(pool, |x| *x as i32).into()),
         },
         DataType::Float => match input {
             Input::FloatTensor(t) => Ok(t.map_in(pool, |x| *x).into()),
             Input::Int32Tensor(t) => Ok(t.map_in(pool, |x| *x as f32).into()),
-            _ => Err(OpError::UnsupportedType),
+            Input::Int8Tensor(t) => Ok(t.map_in(pool, |x| *x as f32).into()),
+            Input::UInt8Tensor(t) => Ok(t.map_in(pool, |x| *x as f32).into()),
+        },
+        DataType::Int8 => match input {
```
Comment (robertknight) [inline src/ops/convert.rs:20]: I think that using `as` for casts is fine for all cases here. Just making some notes for myself:

Converting non-i8 types to i8 will truncate in various ways, and the same for u8. The `Cast` op specs are [here](https://onnx.ai/onnx/operators/onnx__Cast.html). The main cases of interest are:

- Float -> int when out of range: ONNX spec says undefined, so an `as` cast is fine
- Int -> Int when out of range: ONNX spec says `when OOR, discard higher bits and reinterpret (with respect to two’s complement representation for signed types). For example, 200 (int16) -> -56 (int8).`. This matches the behavior of `as` in Rust

File: .DS_Store
Comment (igor-yusupov) [reply .DS_Store:1]: Done. Sorry, I accidentally pushed that file. I also updated .gitignore file.
I think we need to make squash commit to not have this file in git history

File: .gitignore [hunk 2]
```diff
@@ -12,3 +12,5 @@ dist/
 # Converted rten ML models
 *.rten
 models/
+
+*.DS_Store
```
Comment (robertknight) [inline .gitignore:16]: I think there was a misunderstanding. OS/editor-specific ignores shouldn't be part of this _project_'s `.gitignore`. Rather they belong in your personal global gitignore file. See the blog post I linked to in the earlier comment for how to set that up.

Comment (igor-yusupov) [reply .gitignore]: I got it. I have updated global gitignore on my machine and restored the .gitignore file in the repository to its original state. Thank you

**Comment by robertknight** [comment]
Some notes for myself on float-to-int cast compatibility hazards:

The `Cast` [spec](https://onnx.ai/onnx/operators/onnx__Cast.html) says that the behavior of floating point -> integer casts are undefined if out of range. From testing numpy / PyTorch it appears that they have the same behavior as `static_cast<int>(some_float)` in C++.  This behavior can vary [depending on the platform](https://stackoverflow.com/questions/66279679/casting-float-to-int-with-wrap-around-on-aarch64-arm64) and on wasm32 [can trap](https://github.com/WebAssembly/design/issues/986). Rust's `as` on the other hand guarantees saturation at the cost of a potential performance impact. See https://github.com/rust-lang/rust/issues/67058 and https://github.com/rust-lang/rust/issues/10184.

Saturation seems like it would probably be the most useful failure handling and is cross platform. Sticking with `as` seems sensible unless performance turns out to be a major issue.

**Comment by robertknight** [comment]
Note for the changelog: This adds u8 and i8 support to:

- Cast
- Gather
- GatherElements
- GatherND
- ScatterElements
- ScatterND
- Expand
- Flatten
- Reshape
- Squeeze
- Transpose
- Unsqueeze
- Pad

**Pull request merged**